app/src/queue.rs

```diff
@@ -0,0 +1,56 @@
+use {
+    crate::Command,
+    std::pin::Pin,
+    tokio::sync::{broadcast, mpsc},
+};
+
+#[derive(Debug, Clone)]
+pub struct CommandQueue {
+    sender: mpsc::Sender<Command>,
+    killshot: broadcast::Sender<()>,
+}
+
+impl CommandQueue {
+    /// Constructs new [`CommandQueue`] that sends [`Command`]s via `sender`.
+    ///
+    /// In case of a panic of the `sender` channel, a shutdown signal is sent through the
+    /// `killshot`.
+    pub fn new(sender: mpsc::Sender<Command>, killshot: broadcast::Sender<()>) -> Self {
+        Self { sender, killshot }
+    }
+
+    /// Sends a [`Command`] to the background queue for asynchronous processing.
+    pub async fn send(&self, msg: Command) {
+        if self.sender.send(msg).await.is_err() {
+            self.shutdown();
+        }
+    }
+
+    /// Waits for all the commands in the queue to be processed.
+    pub async fn wait_for_pending_commands(&self) {
+        if self
+            .sender
+            .reserve_many(self.sender.max_capacity())
+            .await
+            .is_err()
+        {
+            self.shutdown();
+        }
+    }
+
+    /// Signals the shutdown.
+    pub fn shutdown(&self) {
+        self.killshot
+            .send(())
+            .expect("Sending shutdown request should succeed");
```

Comment:
I think it's fine to just have `.ok()` here. The only way this returns an error is if there are no receivers for the message (i.e. nothing that called `shutdown_listener` is still active), but if there are no receivers then there is nothing to shutdown so the error is not a problem.